x/ibc/06-solomachine/types/msgs.go

```diff
@@ -0,0 +1,182 @@
+	return host.RouterKey
+}
+
+// Type implements sdk.Msg
+func (msg MsgCreateClient) Type() string {
+	return clientexported.TypeMsgCreateClient
+}
+
+// ValidateBasic implements sdk.Msg
+func (msg MsgCreateClient) ValidateBasic() error {
+	if err := msg.ConsensusState.ValidateBasic(); err != nil {
+		return sdkerrors.Wrapf(clienttypes.ErrInvalidConsensus, "consensus state failed validatebasic: %v", err)
+	}
+
+	return host.ClientIdentifierValidator(msg.ClientID)
+}
+
+// GetSignBytes implements sdk.Msg
+func (msg MsgCreateClient) GetSignBytes() []byte {
+	return sdk.MustSortJSON(SubModuleCdc.MustMarshalJSON(msg))
+}
+
+// GetSigners implements sdk.Msg
+func (msg MsgCreateClient) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{sdk.AccAddress(msg.ConsensusState.PubKey.Address())}
+}
+
+// GetClientID implements clientexported.MsgCreateClient
+func (msg MsgCreateClient) GetClientID() string {
+	return msg.ClientID
+}
+
+// GetClientType implements clientexported.MsgCreateClient
+func (msg MsgCreateClient) GetClientType() string {
+	return clientexported.ClientTypeSoloMachine
+}
+
+// GetConsensusState implements clientexported.MsgCreateClient
+func (msg MsgCreateClient) GetConsensusState() clientexported.ConsensusState {
+	return msg.ConsensusState
+}
+
+// MsgUpdateClient defines a message to update an IBC client
+type MsgUpdateClient struct {
+	ClientID string `json:"client_id" yaml:"client_id"`
+	Header   Header `json:"header" yaml:"header"`
+}
+
+// NewMsgUpdateClient creates a new MsgUpdateClient instance
+func NewMsgUpdateClient(id string, header Header) MsgUpdateClient {
+	return MsgUpdateClient{
+		ClientID: id,
+		Header:   header,
+	}
+}
+
+// Route implements sdk.Msg
+func (msg MsgUpdateClient) Route() string {
+	return host.RouterKey
+}
+
+// Type implements sdk.Msg
+func (msg MsgUpdateClient) Type() string {
+	return clientexported.TypeMsgUpdateClient
+}
+
+// ValidateBasic implements sdk.Msg
+func (msg MsgUpdateClient) ValidateBasic() error {
+	if err := msg.Header.ValidateBasic(); err != nil {
+		return sdkerrors.Wrapf(ErrInvalidHeader, "header validatebasic failed: %v", err)
+	}
+	return host.ClientIdentifierValidator(msg.ClientID)
+}
+
+// GetSignBytes implements sdk.Msg
+func (msg MsgUpdateClient) GetSignBytes() []byte {
+	return sdk.MustSortJSON(SubModuleCdc.MustMarshalJSON(msg))
+}
+
+// GetSigners implements sdk.Msg
+func (msg MsgUpdateClient) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{sdk.AccAddress(msg.Header.NewPubKey.Address())}
+}
+
+// GetClientID implements clientexported.MsgUpdateClient
+func (msg MsgUpdateClient) GetClientID() string {
+	return msg.ClientID
+}
+
+// GetHeader implements clientexported.MsgUpdateClient
+func (msg MsgUpdateClient) GetHeader() clientexported.Header {
+	return msg.Header
+}
+
+// MsgSubmitClientMisbehaviour defines an sdk.Msg type that supports submitting
+// Evidence for client misbehaviour.
+type MsgSubmitClientMisbehaviour struct {
```

Comment:
isn't this the same type as the one on ICS07?

Reply:
yea it is. The only sub module specific fields is the `SubModuleCdc`. The rest of the Msg feels like something that could be moved to 02. What do you think I should do for it?

Reply:
Ideally, we should use the message from evidence

Reply:
the message from evidence uses it's own routing key. Besides modifying the message from evidence, we could move the submit client misbehaviour message to 02-client, probably would need to change the cdc registration ordering though

Reply:
I think this is an issue outside of the scope of this pr and should probably just be addressed in a separate pr